Added multiline_latex to more easily print multiline equations

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Inspired by #3001

#### Brief description of what is fixed or changed
A new function `multiline_latex` is added to more easily split multiline equations.

It is simply based on the number of addition/subtraction terms and one can limit the number of terms per line.

Both `align*` and `eqnarray` can be used. I was thinking of adding `IEEEeqnarray` which is better, but then it may not be that regularly used and I couldn't really figure out parameter settings etc to make sense.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
   * New function `multiline_latex` which simplifies generating code for long expressions
<!-- END RELEASE NOTES -->

oscargus committed:
Added multiline_latex to more easily print multiline equations

jksuom commented:
I would prefer having `use_eqnarray` False by default. It seems that the amstex environments produce better spacing.
```
\begin{align*}
  x^2+y^2 &= z^2 \\
  x^3+y^3 &< z^3
\end{align*}

\begin{eqnarray}
  x^2+y^2 &=& z^2 \\
  x^3+y^3 &<& z^3
\end{eqnarray}
```

jksuom commented:
I finally succeeded in uploading an image.

![image](https://user-images.githubusercontent.com/6929119/52479927-ec4aaa00-2bb2-11e9-8687-8200b67284f8.png)

oscargus commented:
I agree with your conclusions in your example. I used beamer and there the conclusion was the opposite:
![image](https://user-images.githubusercontent.com/8114497/52482380-69751f80-2bb1-11e9-86ac-9784b8759086.png)

Now, admittedly, that is not a good reference. I'll try out a few more realistic ones and get back, but that was really the reason I even started to look at something different than `align*` (which was used in the original code).

oscarbenjamin commented:
I thought eqnarray was considered obsolete: [https://tex.stackexchange.com/questions/196/eqnarray-vs-align](https://tex.stackexchange.com/questions/196/eqnarray-vs-align)

oscargus committed:
Made align* default and added IEEEeqnarray

oscargus committed:
Updated documentation examples

oscargus commented:
I wasn't aware that eqnarray has been that deprecated recently...

Anyway, I changed it to align* as default and also added IEEEeqnarray as an option.

oscargus commented on sympy/printing/latex.py:
```diff
@@ -2459,3 +2459,130 @@ def print_latex(expr, **settings):
+
+    environment : "string", optional
+        Which LaTeX wnvironment to use for the output. Options are "align*"
+        (default), "eqnarray", and "IEEEeqnarray". Note the only the first
```
Shuold be "that" instead of "the". Will fix in the next commit, but waiting for other review comments before running a new test...

oscargus commented on sympy/printing/latex.py:
```diff
@@ -2459,3 +2459,130 @@ def print_latex(expr, **settings):
+    terms_per_line : integer, optional
+        Number of terms per line to print. Default is 1.
+
+    environment : "string", optional
```
Gah! Should be `string`, not `"string"`. As below.

oscargus committed:
Merge branch 'master' into termsperline

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -2501,3 +2501,130 @@ def print_latex(expr, **settings):
+
+    # Based on code from https://github.com/sympy/sympy/issues/3001
+    l = LatexPrinter(**settings)
+    if environment.lower().startswith('e'):
```
How about `if environment == 'eqnarray'` (same below)?

oscargus replied:
I was thinking about that, but then one need to type it correctly and considering that the alternatives are `eqnarray`, `IEEEeqnarray`, and `array*` it made sense (to me), to have a more "sloppy" checking of alternatives so that `array` and `ieeeeqnarray` also works.

oscargus replied:
Clearly, I mean `align*`, not `array*`

oscarbenjamin replied:
I don't see the advantage of sloppy checking. It's better to list the possibilities in the docstring then someone can get it right first time they write their code. Allowing anything that starts with e here leads to a backward compatibility problem if you ever decide to add another environment that begins with e.

oscargus replied:
OK! Should it be case sensitive or not?

oscarbenjamin replied:
I would go case-sensitive. I can see that you have already listed the 3 possibilities in the docstring so I'd go with that. Also these are the names of LaTeX environments which are case-sensitive.

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -2501,3 +2501,130 @@ def print_latex(expr, **settings):
+        nonumber = r'\nonumber'
+        end_term = '\n\\end{IEEEeqnarray}'
+        doubleet = True
+    else:
```
Maybe the else could be an error after all correct cases are handled

oscargus replied:
Yes, that may be an idea.

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -2501,3 +2501,130 @@ def print_latex(expr, **settings):
+        term = terms[i]
+        term_start = ''
+        term_end = r''
+        sign = r'+'
```
I don't understand why the above are raw strings (no big deal...)

oscargus replied:
Probably copy-paste or that they initially included things where it was needed... Makes sense to remove it.

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -2501,3 +2501,130 @@ def print_latex(expr, **settings):
+    \end{eqnarray}
+
+    Using "IEEEeqnarray":
+    \begin{IEEEeqnarray}{rCl}
```
Is there a missing print here?

oscargus replied:
Yes. Thanks!

oscarbenjamin reviewed (commented):
I left a couple of comments. Your own comments also seemed reasonable

oscargus commented:
Thanks for the review @oscarbenjamin ! I'll update it eventually.

oscargus committed:
Updated based on review comments

oscargus committed:
Merge branch 'termsperline' of https://github.com/oscargus/sympy into termsperline

oscargus committed:
Fixed failing doctest

oscargus commented:
Comments fixed (apart from the sloppy environment detection).

oscargus committed:
Using complete environment names as per request

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -2501,3 +2501,130 @@ def print_latex(expr, **settings):
+    & & - \cos{\left(\log{\left(y \right)} \right)}
+    \end{eqnarray}
+
+    Using "IEEEeqnarray":
```
Should this be double quotes or backticks like above?

oscarbenjamin commented on sympy/printing/latex.py:
```diff
@@ -2501,3 +2501,130 @@ def print_latex(expr, **settings):
+def multiline_latex(lhs, rhs, terms_per_line=1, environment="align", use_dots=False, **settings):
+    r"""
+    This function generates a LaTeX equation with a multiline right-hand side
+    in either an ``eqnarray`` or ``align*`` environment.
```
Below you list 3 environments including IEEEeqnarray but here only the other two.

oscarbenjamin reviewed (commented):
I've left comments but otherwise good.

oscargus commented:
Good comments! Will fix.

oscargus committed:
Merge branch 'master' into termsperline

oscargus committed:
Fixed review comments

oscargus committed:
Merge with master, resolved conflicts

oscargus commented:
I have merged with current master (fixes done since long, but didn't comment on that).

oscarbenjamin commented:
+1 to merge if you're finished with it.

oscargus commented:
OK! I merge it then!

moorepants commented:
Is there a way to enable this with `init_printing()`?

moorepants commented:
I tried `init_printing(latex_printer=multiline_latex)` naively, but that fails with:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/miniconda/envs/multibody-book/lib/python3.10/site-packages/IPython/core/formatters.py:343, in BaseFormatter.__call__(self, obj)
    341     method = get_real_method(obj, self.print_method)
    342     if method is not None:
--> 343         return method()
    344     return None
    345 else:

File ~/miniconda/envs/multibody-book/lib/python3.10/site-packages/sympy/interactive/printing.py:226, in _init_ipython_printing.<locals>._print_latex_text(o)
    222 """
    223 A function to generate the latex representation of sympy expressions.
    224 """
    225 if _can_print(o):
--> 226     s = latex(o, mode=latex_mode, **settings)
    227     if latex_mode == 'plain':
    228         return '$\\displaystyle %s$' % s

TypeError: multiline_latex() missing 1 required positional argument: 'rhs'
```
